### PR TITLE
Implement `MsgFindAndModify` for Tigris

### LINKED ISSUE #775: Implement `MsgFindAndModify` for Tigris
We assume that we have `tigrisdb` package in some form already.

It should be as simple as copying and pasting PostgreSQL handler's code and tweaking it. We should also change `tigrisdb` to look more similar to `pgdb`, with one big exception of QueryDocuments – we should not use channel-based API for that one.

New compatibility tests for `findAndModify` are in the scope.

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

This PR closes #775.

As part of the PR `findAndModify` tests were moved to `findAndModifyCompat` tests when possible.

## Other changes

- `PrepareFindAndModifyParams` is moved to the `common` handler as it was easy. In the future, we might want to move more things to the common handler, but for now, such refactoring would take some time, so it's not in the scope of this task. 
- `FetchParam` and `SQLParam` are passed as pointers everywhere.

## Readiness checklist

* [ ] I added tests for new functionality or bugfixes.
* [ ] I ran `task all`, and it passed.
* [ ] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [ ] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handlers/pg/msg_count.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func (h *Handler) MsgCount(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, e
 		}
 	}
 
-	var sp pgdb.SQLParam
+	sp := new(pgdb.SQLParam)
```
Comment (noisersup) [inline internal/handlers/pg/msg_count.go:64]: I don't understand why this change is needed.

Comment (rumyantseva) [reply internal/handlers/pg/msg_count.go:64]: @AlekSi suggested using pointers for `SQLParam` in one of the functions, so I think we should move to pointers everywhere (for consistency). 

@noisersup if you have an opinion about it (where we should use values and where pointers) please share.

Comment (noisersup) [reply internal/handlers/pg/msg_count.go:64]: So I assume that's for using a little bit less memory? :)

Comment (rumyantseva) [reply internal/handlers/pg/msg_count.go:64]: I think so. To be honest, to me, in this particular case the difference is not that huge. I can agree that if all the four fields of `SQLParam` are set, we can win some memory by passing by the pointer.

File: internal/handlers/common/msg_findandmodify.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+	if returnNewDocument, err = GetBoolOptionalParam(document, "new"); err != nil {
+		return nil, err
+	}
+
+	var upsert bool
+
+	if upsert, err = GetBoolOptionalParam(document, "upsert"); err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:72]: Just to make spacing a bit more consistent. 😃 
```suggestion
	var db string
	if db, err = GetRequiredParam[string](document, "$db"); err != nil {
		return nil, err
	}

	var collection string
	if collection, err = GetRequiredParam[string](document, command); err != nil {
		return nil, err
	}

	if collection == "" {
		return nil, NewErrorMsg(
			ErrInvalidNamespace,
			fmt.Sprintf("Invalid namespace specified '%s.'", db),
		)
	}

	var remove bool
	if remove, err = GetBoolOptionalParam(document, "remove"); err != nil {
		return nil, err
	}

	var returnNewDocument bool
	if returnNewDocument, err = GetBoolOptionalParam(document, "new"); err != nil {
		return nil, err
	}

	var upsert bool
	if upsert, err = GetBoolOptionalParam(document, "upsert"); err != nil {
		return nil, err
	}
```

**Review by w84thesun — CHANGES_REQUESTED** [review]
Overall looks good, small suggestion about spaces inside `PrepareFindAndModifyParams`.

**Review by w84thesun — DISMISSED** [review]

**Review by noisersup — APPROVED** [review]

**Review by noisersup — CHANGES_REQUESTED** [review]
Tests are failing after merge

File: internal/handlers/pg/msg_count.go (hunk 1 above)
Comment (AlekSi) [reply internal/handlers/pg/msg_count.go:64]: FWIW, I think we should standardize on
```go
var x T
foo(&x)
```

See https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_use_a_consistent_declaration_style

**Review by noisersup — COMMENTED** [review]
LGTM, but for consistency I agree with Aleksi in the topic of `SQLParam` declarations.

**Review by w84thesun — APPROVED** [review]

File: internal/handlers/pg/msg_count.go (hunk 1 above)
Comment (rumyantseva) [reply internal/handlers/pg/msg_count.go]: Fixed!

**Review by noisersup — APPROVED** [review]
:+1:

File: internal/handlers/common/msg_findandmodify.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+func PrepareFindAndModifyParams(document *types.Document) (*FindAndModifyParams, error) {
+	var err error
+
+	command := document.Command()
+
+	var db string
+
+	if db, err = GetRequiredParam[string](document, "$db"); err != nil {
+		return nil, err
+	}
```
Comment (AlekSi) [inline internal/handlers/common/msg_findandmodify.go:43]:
```suggestion
	db, err := GetRequiredParam[string](document, "$db")
	if err != nil {
		return nil, err
	}
```

I know `wsl` could be annoying, but we should not follow it blindly

File: integration/shareddata/composites.go [hunk 4]
```diff
@@ -48,6 +48,15 @@ var Composites = &Values[string]{
 	},
 }
 
+// ArrayInt32s contains int32 arrays for tests.
+var ArrayInt32s = &Values[string]{
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply integration/shareddata/composites.go:52]: Yeah, I'll remove it for now (wanted to add some cases for arrays, but it doesn't make sense as arrays are not implemented yet).

**Comment by rumyantseva** [comment]
The PR is ready for review. Feel free to suggest and merge improvements if any.

File: internal/handlers/common/msg_findandmodify.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+	if returnNewDocument, err = GetBoolOptionalParam(document, "new"); err != nil {
+		return nil, err
+	}
+
+	var upsert bool
+
+	if upsert, err = GetBoolOptionalParam(document, "upsert"); err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:73]:
```suggestion
	var remove bool
	if remove, err = GetBoolOptionalParam(document, "remove"); err != nil {
		return nil, err
	}

	var returnNewDocument bool
	if returnNewDocument, err = GetBoolOptionalParam(document, "new"); err != nil {
		return nil, err
	}

	var upsert bool
	if upsert, err = GetBoolOptionalParam(document, "upsert"); err != nil {
		return nil, err
	}
```

File: internal/handlers/common/msg_findandmodify.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	db, err := GetRequiredParam[string](document, "$db")
+	if err != nil {
+		return nil, err
+	}
+
+	var collection string
+
+	if collection, err = GetRequiredParam[string](document, command); err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:48]:
```suggestion
	var collection string
	if collection, err = GetRequiredParam[string](document, command); err != nil {
		return nil, err
	}
```

File: internal/handlers/common/msg_findandmodify.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+// PrepareFindAndModifyParams prepares findAndModify request fields.
+func PrepareFindAndModifyParams(document *types.Document) (*FindAndModifyParams, error) {
+	var err error
+
+	command := document.Command()
+
+	db, err := GetRequiredParam[string](document, "$db")
+	if err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:42]:
```suggestion
	var db string
	if db, err = GetRequiredParam[string](document, "$db"); err != nil {
		return nil, err
	}
```

File: internal/handlers/common/msg_findandmodify.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+	var sort *types.Document
+	if sort, err = GetOptionalParam(document, "sort", sort); err != nil {
+		return nil, err
+	}
+
+	maxTimeMS, err := GetOptionalPositiveNumber(document, "maxTimeMS")
+	if err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:88]:
```suggestion
	var maxTimeMS int32
	if maxTimeMS, err = GetOptionalPositiveNumber(document, "maxTimeMS"); err != nil {
		return nil, err
	}
```

**Review by w84thesun — CHANGES_REQUESTED** [review]

File: internal/handlers/common/msg_findandmodify.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+	var sort *types.Document
+	if sort, err = GetOptionalParam(document, "sort", sort); err != nil {
+		return nil, err
+	}
+
+	var maxTimeMS int32
+	if maxTimeMS, err = GetOptionalPositiveNumber(document, "maxTimeMS"); err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:84]:
```suggestion
	db, err := GetRequiredParam[string](document, "$db")
	iff err != nil {
		return nil, err
	}

	collection, err := GetRequiredParam[string](document, command)
	if err != nil {
		return nil, err
	}

	if collection == "" {
		return nil, NewErrorMsg(
			ErrInvalidNamespace,
			fmt.Sprintf("Invalid namespace specified '%s.'", db),
		)
	}

	remove, err := GetBoolOptionalParam(document, "remove")
	if err != nil {
		return nil, err
	}

	returnNewDocument, err := GetBoolOptionalParam(document, "new")
	if err != nil {
		return nil, err
	}

	upsert, err := GetBoolOptionalParam(document, "upsert")
	if err != nil {
		return nil, err
	}

	query, err := GetOptionalParam(document, "query", query)
	if err != nil {
		return nil, err
	}

	sort, err := GetOptionalParam(document, "sort", sort)
	if err != nil {
		return nil, err
	}

	maxTimeMS, err := GetOptionalPositiveNumber(document, "maxTimeMS")
	if err != nil {
		return nil, err
	}
```

**Review by w84thesun — CHANGES_REQUESTED** [review]

File: internal/handlers/common/msg_findandmodify.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+	query, err := GetOptionalParam(document, "query", query)
+	if err != nil {
+		return nil, err
+	}
+
+	sort, err := GetOptionalParam(document, "sort", sort)
+	if err != nil {
+		return nil, err
+	}
```
Comment (w84thesun) [inline internal/handlers/common/msg_findandmodify.go:79]:
```suggestion
	query, err := GetOptionalParam(document, "query", new(types.Document))
	if err != nil {
		return nil, err
	}

	sort, err := GetOptionalParam(document, "sort", new(types.Document))
	if err != nil {
		return nil, err
	}
```

**Review by w84thesun — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]
